src/webgl/p5.RendererGL.js

```diff
@@ -879,29 +879,44 @@ p5.RendererGL = class RendererGL extends p5.Renderer {
     this.curStrokeJoin = join;
   }
 
-  filter(...args) {
-    // Couldn't create filterFramebuffer in RendererGL constructor
-    // (.framebuffers was still undefined)
+  filter(userShader) {
+    // Couldn't create graphics in RendererGL constructor
+    // (led to infinite loop)
     // so it's just created here once on the initial filter call.
-    if (!this.filterFramebuffer) {
-      this.filterFramebuffer = new p5.Framebuffer(this._pInst);
+    if (!this.filterGraphicsLayer) {
+      this.filterGraphicsLayer =
+        new p5.Graphics(this.width, this.height, constants.WEBGL, this._pInst);
     }
 
-    // TODO:
+    // TODO, handle different signatures:
     // if args[0] is a p5.Shader,
     //   this.filterShader = args[0]
     // else
-    //   this.filterShader = selectFromShaderFilterDefaults(args[0])
-    //   filterOperation = undefined or args[1]
-    //
-    // this.filterFramebuffer.begin();
-    //
-    // apply shader
-    // transfer pixels to main canvas
-    //
-    // this.filterFramebuffer.end();
-
-    p5._friendlyError('webgl filter implementation in progress');
+    //   this.filterShader = map(args[0], {GRAYSCALE: grayscaleShader, ...})
+    //   filterOperationParameter = undefined or args[1]
+
+    let pg = this.filterGraphicsLayer;
+
+    // not sure why userShader[0] is the actual shader object
+    // and userShader is Arguments instead
+    userShader = userShader[0];
+    // perhaps necessary: bind the shader to the pg renderer, not the main
+    // let userShaderCopy =
+    //   new p5.Shader(pg._renderer, userShader._vertSrc, userShader._fragSrc);
+
+    // apply shader to pg
+    // TODO: shader isn't being applied properly
+    pg.shader(userShader);
```

Comment:
I think user shader will need to be created using the pg context here. I wonder if we can just extract the shader strings from an already loaded shader and reload it using `pg.createShader()`